### PR TITLE
ENH: stats: add one-sample Monte Carlo hypothesis test

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
gh-13371 - bootstrapping
gh-13899 - permutation test

#### What does this implement/fix?
To fill in our offerings of fundamental resampling methods, this adds Monte Carlo hypothesis tests for assessing whether a sample was drawn from a given distribution. It can reproduce the results of hypothesis tests like `ks_1samp`, `skewtest`, `kurtosistest`, `normal test`, `cramervonmises`, and `anderson`, but avoids small sample size limitations.

#### Additional information
- [x] ~~What should the function be named?~~ _Update: `monte_carlo_test`_
- [x] ~~This could be probably be generalized, or it could even be worked into `permutation_test` - but would that be useful?~~  _Update: This could at least be generalized to the multivariate case, but maybe in a second PR._

~~I'm still working on the requirements for the provided `rvs` and whether this function should also accept a random state object or whether `rvs` should take care of its own state without input.~~ `rvs` should take care of its own state.

~~Let's work on gh-13899 before diving into this one.~~

### CODE REVIEW

File: scipy/stats/__init__.py [hunk 1]
```diff
@@ -329,6 +328,15 @@
...
    stats.qmc
 
+Resampling Methods
+==================
+
+.. autosummary::
+   :toctree: generated/
+
+   bootstrap
+   permutation_test
```
Comment (mdhaber) [inline scipy/stats/__init__.py:338]:
```suggestion
```

**Comment by mdhaber** [comment]
@chrisb83 Thanks for reviewing `permutation_test`! Do you think you'd be interested in this one at some point? If so, I'm looking for suggestions about whether (and if so, how) I should generalize it.

**Comment by mdhaber** [comment]
@tirthasheshpatel does this interest you?

**Comment by tirthasheshpatel** [comment]
> does this interest you?

Yes! I can help with this.

File: scipy/stats/_bootstrap.py [hunk 2]
```diff
@@ -486,3 +486,228 @@ def percentile_fun(a, q):
...
+    0.701070107010701
+
+    The probability of obtaining a test statistic less than or equal to the
+    observed value under the null hypothesis is 67.77%. This is greater than
+    our chosen threshold of 5%, so we cannot consider this to to be significant
+    evidence against the null hypothesis.
+
+    Note that the this p-value essentially matches that of `stats.skewtest`,
+    which relies on an asymptotic distribution of a test statistic based on
+    the sample skewness.
```
Comment (mdhaber) [inline scipy/stats/_bootstrap.py:650]:
```suggestion
    Note that this p-value essentially matches that of
    `scipy.stats.skewtest`, which relies on an asymptotic distribution of a
    test statistic based on the sample skewness.
```

File: scipy/stats/_bootstrap.py [hunk 3]
```diff
@@ -486,3 +486,228 @@ def percentile_fun(a, q):
...
+    >>> # because our statistic is vectorized, we pass `vectorized=True`
+    >>> rvs = lambda size: stats.norm.rvs(size=size, random_state=rng)
+    >>> res = monte_carlo_test(x, rvs, statistic, vectorized=True)
+    >>> print(res.statistic)
+    0.12457412450240658
+    >>> print(res.pvalue)
+    0.701070107010701
+
+    The probability of obtaining a test statistic less than or equal to the
+    observed value under the null hypothesis is 67.77%. This is greater than
```
Comment (chrisb83) [inline scipy/stats/_bootstrap.py:644]: rather: ~70%

File: scipy/stats/_bootstrap.py [hunk 4]
```diff
@@ -486,3 +486,228 @@ def percentile_fun(a, q):
...
+
+    >>> stats.skewtest(x).pvalue
+    0.6892046027110614
+
+    This asymptotic approximation is not valid for small sample sizes, but
+    `monte_carlo_test` can be used with samples of any size.
+
+    >>> x = stats.skewnorm.rvs(a=1, size=7, random_state=rng)
+    >>> # stats.skewtest(x) would produce an error due to small sample
+    >>> res = monte_carlo_test(x, rvs, statistic, vectorized=True)  # OK
```
Comment (chrisb83) [inline scipy/stats/_bootstrap.py:660]: remove "# OK"

File: scipy/stats/tests/test_bootstrap.py [hunk 5]
```diff
@@ -470,3 +472,230 @@ def statistic_1d(*data):
...
+
+        message = "`alternative` must be in..."
+        with pytest.raises(ValueError, match=message):
+            monte_carlo_test([1, 2, 3], stats.norm.rvs, stat,
+                             alternative='ekki')
+
+    def test_batch(self):
+        # make sure that the `batch` parameter is respected by checking the
+        # maximum batch size provided in calls to `statistic`
+        np.random.seed(0)
```
Comment (chrisb83) [inline scipy/stats/tests/test_bootstrap.py:530]: I think Pamphile tried to get rid of the legacy approach using RandomState. consider replacing it by `rng = np.random.default_rng(...)` and `x = rng.random(10)`

**Review by chrisb83 — COMMENTED** [review]
Some initial comments after reading the documentation and code

File: scipy/stats/tests/test_bootstrap.py (hunk 5 above)
Comment (mdhaber) [reply scipy/stats/tests/test_bootstrap.py:530]: Yes, yes... : P It's sort of a game now where I try to slide this thing under his radar, so if you don't personally care, I won't tell.

**Comment by mdhaber** [comment]
Thanks for taking a look @chrisb83!

File: scipy/stats/_bootstrap.py (hunk 4 above)
Comment (mdhaber) [reply scipy/stats/_bootstrap.py:660]:
```suggestion
    >>> res = monte_carlo_test(x, rvs, statistic, vectorized=True)
```

File: scipy/stats/tests/test_bootstrap.py (hunk 5 above)
Comment (tupui) [reply scipy/stats/tests/test_bootstrap.py:530]: 🙈

File: scipy/stats/_bootstrap.py [hunk 6]
```diff
@@ -486,3 +486,228 @@ def percentile_fun(a, q):
...
+    >>> res = monte_carlo_test(x, rvs, statistic, vectorized=True)
+
+    The Monte Carlo distribution of the test statistic is provided for
+    further investigation.
+
+    >>> import matplotlib.pyplot as plt
+    >>> plt.hist(res.null_distribution, bins=50)
+    >>> plt.title("Monte Carlo distribution of test statistic")
+    >>> plt.xlabel("Value of Statistic")
+    >>> plt.ylabel("Frequency")
```
Comment (tupui) [inline scipy/stats/_bootstrap.py:669]: We have to use the OOP version of MPL for new code.
```suggestion
    >>> import matplotlib.pyplot as plt
    >>> fig, ax = plt.subplots()
    >>> ax.hist(res.null_distribution, bins=50)
    >>> ax.set_title("Monte Carlo distribution of test statistic")
    >>> ax.set_xlabel("Value of Statistic")
    >>> ax.set_ylabel("Frequency")
    >>> plt.show()
```

**Review by tupui — COMMENTED** [review]
Thanks @mdhaber. Small remarks.

For later, it would be nice to have a tutorial or quick guidelines on what to use and when since we have a few methods now.

File: scipy/stats/_bootstrap.py [hunk 7]
```diff
@@ -486,3 +486,228 @@ def percentile_fun(a, q):
...
+
+    >>> import numpy as np
+    >>> from scipy import stats
+    >>> def statistic(x, axis):
+    ...     return stats.skew(x, axis)
+
+    After collecting our data, we calculate the observed value of the test
+    statistic.
+
+    >>> rng = np.random.default_rng(1638083107694713882823079058616272161)
```
[the remark this answers is not in the document]
Comment (mdhaber) [reply scipy/stats/_bootstrap.py:622]: Probably OK to do it now.
```suggestion
    >>> rng = np.random.default_rng()
```

**Comment by mdhaber** [comment]
I was looking at gh-15463 and noticed that `monte_carlo_test` function could be useful for that if we add support for multivariate distributions. However, because the function is vectorized to process multiple samples at the same time, we have a bit of a challenge: besides knowing which axis the statistic is taken over, we need to know which axis/axes correspond with the dimensions of observations from the distribution.

For instance, the user were to pass in a `sample` with shape `3, 100, 4`, `rvs` ~ `multivariate_normal.rvs` (wrapped so that `mean` and `cov` are set), and `axis = 1`, we don't necessarily know whether this is is to be treated as 3 separate samples from a 4-dimensional normal distribution or 4 samples from a 3-dimensional normal distribution. Likewise, the user's `statistic` function doesn't know which axes are which when I pass it Monte Carlo samples from the distribution.

A few obvious solutions come to mind:
* Insist (in documentation) that dimensions of _individual_ observations of the distribution always come last. With this convention, the example above is for 3 separate samples from a 4-dimensional normal distribution. The convention is reasonable because when we call `multivariate_normal.rvs(np.zeros(3), np.eye(3), size = 4)`, we get a result with shape `(4, 3)`. We will be able to tell whether individual observations need 0, 1, or 2 (or more) axes when we call the `rvs` method (we can compare the number of dimensions in the result against `len(size)` that we pass in). When we pass the `statistic` function random variates, we'll trust that `rvs` has followed this convention when generating the random variates. (A slight modification to the above might be possible - I think that we don't have to insist on a particular convention as long as the user provides a `statistic` function that is compatible with its `rvs` function, but I'd have to think more about this.)
* Add an explicit, second axis parameter (named ?) that specifies which dimension(s) correspond with those of a single observation from the distribution. For instance, if this second axis parameter were 0, then the example above is actually 4 separate samples from a 3-dimensional normal distribution.

Thoughts? Or for simplicity, we could leave extension to multivariate distributions to a followup.

**Comment by tupui** [comment]
> Thoughts? Or for simplicity, we could leave extension to multivariate distributions to a followup.

I think we can just leave this to a follow up. We don't have much multivariate distributions, so I would not spend to much time on this.

**Review by tupui — APPROVED** [review]
Looks great to me, the interface is clean and it works good. If we want to add more things we can do it later (`workers` would be nice). Merging since CI is green. Thanks again @mdhaber 

p.s. yes love the RNG, but use of 9999 got my heart 😉

**Comment by tupui** [comment]
@mdhaber It makes sense to group them 👍 Not sure about the name though, but it's private so we can always revisit.

**Pull request merged**